server/src/users/__mocks__/users.service.ts

```diff
@@ -0,0 +1,6 @@
+import { userEntityStub } from '../../../test/stubs';
+
+// https://medium.com/@deanslamajr/jest-fn-all-the-things-d26f3b929986
+export const UsersService = jest.fn().mockReturnValue({
+  create: jest.fn().mockReturnValue(userEntityStub()),
+});
```

Comment:
in this particular test, we are asserting that the argument passed (not the returned type) to the service is the  `userCreateDtoStub`.

Reply:
> i think jest will still find these mocks if we put them in the test folder so it'd be `src/test/__mocks__/...` advantage here would be all mocks in 1 place, easier to copy new ones also down the line we might need services in other controllers (i.e. need users service in assets controller) so maybe one place in a higher up folder makes more sense rather than local, which would either cause long import paths or repeated mocks in all the different folders

Makes perfect sense. Although, if we move it to the `test` dir, [we loose the `auto-inject` utility ](https://jestjs.io/docs/manual-mocks#mocking-user-modules)an will have to add a little more code - something like:

```typescript
import { UsersService as UsersServiceMock } from './__mocks__';

// jest.mock('./users.service');

describe('UsersController', () => {
  let controller: UsersController;
  let usersService: UsersService = new UsersServiceMock();

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      controllers: [UsersController],
      providers: [{ provide: UsersService, useValue: usersService }],
    }).compile();
```

Reply:
ah you're right, thought if we moved it jest would still be able to auto-inject but docs say `subdirectory immediately adjacent to the module`

Reply:
Let me give it a try

Reply:
> yeah looking at it now, i'm not sure the tests you wrote are using this mock in this mock you have create returning `userEntityStub` but in the tests you expect `userCreateDtoStub`

so expanding off this...
i'm not sure if this is the ideal mock for this
this will always return user A
but if we pass in user B we want to get back user B (maybe not effecting current tests but i expect it'll be a thing in the future)
so should this mock be something more like `create: jest.fn(newUser => { delete newUser.password; return newUser; }),`

Reply:
> ah you're right, thought if we moved it jest would still be able to auto-inject but docs say `subdirectory immediately adjacent to the module`

After testing, Jest auto inject doesn't work when `__mocks__` is in a higher directory, I think we'll be fine giving up auto-inject over having all mocks in one test dir. Manually injecting into the testing module won't be that difficult.

Reply:
> > yeah looking at it now, i'm not sure the tests you wrote are using this mock in this mock you have create returning `userEntityStub` but in the tests you expect `userCreateDtoStub`
> 
> so expanding off this... i'm not sure if this is the ideal mock for this this will always return user A but if we pass in user B we want to get back user B (maybe not effecting current tests but i expect it'll be a thing in the future) so should this mock be something more like `create: jest.fn(newUser => { delete newUser.password; return newUser; }),`

^ this sounds like a good test case for the service

Reply:
marking these as resolved, because PR is now shifting to integration tests